plugins/gateway/kong/plugin.go

```diff
@@ -29,39 +29,71 @@ import (
 	"github.com/apiclarity/apiclarity/plugins/api/client/client/operations"
 	"github.com/apiclarity/apiclarity/plugins/api/client/models"
 	"github.com/apiclarity/apiclarity/plugins/common"
+	"github.com/apiclarity/apiclarity/plugins/common/tsm"
 )
 
+var traceSamplingManager *tsm.TraceSamplingManager = nil
+var telemetriesAPI *client.APIClarityPluginsTelemetriesAPI = nil
+
 type Config struct {
-	EnableTLS bool   `json:"enable_tls"`
-	Host      string `json:"host"`
-	apiClient *client.APIClarityPluginsTelemetriesAPI
+	EnableTLS            bool   `json:"enable_tls"`
+	Host                 string `json:"host"`
+	TraceSamplingHost    string `json:"trace_sampling_host"`
+	TraceSamplingEnabled bool   `json:"trace_sampling_enabled"`
 }
 
 func New() interface{} {
 	return &Config{}
 }
 
 func (conf Config) Access(kong *pdk.PDK) {
+	if conf.TraceSamplingEnabled && traceSamplingManager == nil {
+		_ = kong.Log.Info("creating trace sampling manager")
+		traceSampling, err := tsm.Create(conf.EnableTLS, conf.TraceSamplingHost)
+		if err != nil {
+			_ = kong.Log.Err(fmt.Sprintf("Failed to create trace sampling manager: %v", err))
+		} else {
+			traceSamplingManager = traceSampling
+			traceSamplingManager.Start()
+		}
+	}
+	if conf.TraceSamplingEnabled {
+		if err := setIgnoredHostOnSharedContext(kong); err != nil {
+			_ = kong.Log.Err(fmt.Sprintf("Failed to set ignore host on shared context: %v", err))
+		}
+	}
+
+	// set request time on shared context
 	if err := kong.Ctx.SetShared(common.RequestTimeContextKey, time.Now().UTC().UnixMilli()); err != nil {
 		_ = kong.Log.Err(fmt.Sprintf("Failed to set request time on shared context: %v", err))
 	}
 }
 
 func (conf Config) Response(kong *pdk.PDK) {
 	_ = kong.Log.Info("Handling telemetry")
-	if conf.apiClient == nil {
+	if conf.TraceSamplingEnabled {
+		shouldIgnore, err := getShouldIgnoreHostFromContext(kong)
+		if err != nil {
+			_ = kong.Log.Err(fmt.Sprintf("Failed to get should ignore host from context: %v", err))
+		}
+		if shouldIgnore == "true" {
+			_ = kong.Log.Info("Ignoring host")
+			return
```

Comment:
can't we run `TraceSamplingManager.ShouldTrace(host)` here and avoid writing/reading to context?